### PR TITLE
BUG: np.asscalar should pass through scalars

### LINKED ISSUE #4701: np.asscalar should pass through scalars
**Suggestion**
It would be very convenient if `np.asscalar()` would pass through scalars.  The use-case is to ensure that an argument is a scalar, regardless what is passed in (as long as it behaves as a scalar).

``` python
>>> np.asarray(1.0)
1.0
>>> np.asarray(1j)
1j
```

Presently these fail as the implementation simply looks for the `.item()` method, which scalars do not possess:

`````` python
>>> np.asarray(1.0)
AttributeError                            Traceback (most recent call last)
<ipython-input-13-c6bc08be135e> in <module>()
----> 1 np.asscalar(1.0)

.../python2.7/site-packages/numpy/lib/type_check.pyc in asscalar(a)
    461 
    462     """
--> 463     return a.item()
    464 
    465 #-----------------------------------------------------------------------------

AttributeError: 'float' object has no attribute 'item'```
``````

**Rational**
It is common to use `np.asarray()` to ensure that the argument is an array (in the case that it might be a list etc.).  In an analogous use case, it would be useful to be able to use `np.asscalar()` to ensure that an argument is a scalar (as long as it can be converted to a scalar).

**Workaround**
Convert the argument to an array first.

``` python
np.asscalar(np.asarray(x))
```

**Resolution**
Perhaps code like the following would work:

``` python
def asscalar(a):
    try:
        return a.item()
    except AttributeError, e:
        return np.asarray(a).item()
```

### PR DESCRIPTION
**Pull request by seawolf631**
Closes #4701

### CODE REVIEW

File: numpy/lib/type_check.py [hunk 1]
```diff
@@ -479,7 +479,10 @@ def asscalar(a):
     24
 
     """
-    return a.item()
+    try:
+        return a.item()
+    except AttributeError, e:
```
Comment (eric-wieser) [inline numpy/lib/type_check.py]: Deprecated syntax, just use `except AttributeError`

**Comment by eric-wieser** [comment]
Right now this just duplicates #5126 - so make sure to address the comments there

**Comment by seawolf631** [comment]
My new changes pass through all scalars, and then for arrays of size greater than 1 you get a ValueError of "Array needs to be of size 1".

**Comment by eric-wieser** [comment]
This changed the behavior of `np.asscalar(np.float64(1))`. Before it returned `float`, now it returns `float64`.

This function strikes me as one in the "mistake we're now stuck with" category, which might be best to just leave alone. Calling `item()` in general is a pretty bad idea.

**Pull request closed without merging**